chef/cookbooks/monasca/recipes/master.rb

```diff
@@ -0,0 +1,96 @@
+#
+# Copyright 2017 Fujitsu LIMITED
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitation.
+
+package "ansible"
+
+keystone_settings = KeystoneHelper.keystone_settings(node, @cookbook_name)
+
+directory "/opt/fujitsu" do
+  owner "root"
+  group "root"
+  mode "0755"
+end
+
+# TODO: use package "monasca-installer"
+remote_file "/opt/fujitsu/cmm-suse.tgz" do
+  source node[:monasca][:master][:cmm_tarball_url]
+  owner "root"
+  group "root"
+  mode "0644"
+  action :create_if_missing
+end
+
+execute "extract cmm tarball" do
+  command "tar xf cmm-suse.tgz"
+  cwd "/opt/fujitsu/"
+  not_if { Dir.exist?("/opt/fujitsu/monasca-installer") }
+  notifies :run, "execute[run ansible]", :delayed
+end
+
+template "/opt/fujitsu/monasca-installer/credentials.yml" do
+  source "credentials.yml.erb"
+  owner "root"
+  group "root"
+  mode "0644"
+  variables(
+    keystone_settings: keystone_settings
+  )
+  notifies :run, "execute[run ansible]", :delayed
+end
+
+template "/opt/fujitsu/monasca-installer/cmm-hosts" do
+  source "cmm-hosts.erb"
+  owner "root"
+  group "root"
+  mode "0644"
+  variables(
+    # TODO: use all monasca-server hosts
+    monasca_host: MonascaHelper.monasca_hosts(search(:node, "roles:monasca-server"))[0],
+    ssh_user_monasca: "root",
+    offline_resources_host: CrowbarHelper.get_host_for_admin_url(node),
+    ssh_user_offline_host: "root",
+    keystone_host: keystone_settings["public_url_host"],
+    ssh_user_keystone: "root"
+  )
+  notifies :run, "execute[run ansible]", :delayed
+end
+
+template "/opt/fujitsu/monasca-installer/group_vars/all_group" do
+  source "all_group.erb"
+  owner "root"
+  group "root"
+  mode "0644"
+  variables(
+    keystone_settings: keystone_settings
+  )
+  notifies :run, "execute[run ansible]", :delayed
+end
+
+file "/opt/fujitsu/monasca-installer/.installed" do
```

Comment:
Yes, definitely document this in detail. Also, I can't think of a better way to make sure that exec resource gets retried either. This is tricky with exec, because it doesn't have any state for chef to check.